streams/shared/src/main/scala/zio/stream/Structure.scala

```diff
@@ -0,0 +1,148 @@
+    Concat(self, () => other)
+
+  def map[B](f: A => B): Structure[R, E, B] =
+    self match {
+      case Iterator(iter)          => Structure.Iterator(iter.map(_.map(f)))
+      case Concat(hd, tl)          => Structure.Concat(hd.map(f), () => tl().map(f))
+      case s: FlatMap[R, E, a0, A] => Structure.FlatMap(s.fa, (el: a0) => s.f(el).map(f))
+      case Effect(zio)             => Structure.Effect(zio.map(f))
+    }
+
+  def mapM[R1 <: R, E1 >: E, B](f: A => ZIO[R1, E1, B]): Structure[R1, E1, B] =
+    self match {
+      case Concat(hd, tl)          => Concat(hd.mapM(f), () => tl().mapM(f))
+      case Iterator(iter)          => Iterator(iter.map(_.flatMap(f(_).mapError(Some(_)))))
+      case s: FlatMap[R, E, a0, A] => Structure.FlatMap(s.fa, (el: a0) => s.f(el).mapM(f))
+      case Effect(zio)             => Structure.Effect(zio.flatMap(f))
+    }
+
+  def flatMap[R1 <: R, E1 >: E, B](f: A => Structure[R1, E1, B]): Structure[R1, E1, B] =
+    FlatMap[R1, E1, A, B](self, f)
+
+  def process: ZManaged[R, E, Pull[R, E, A]] =
+    self match {
+      case Iterator(iter) => iter
+      case _ =>
+        for {
+          refs <- (Ref.make[Structure[R, E, _]](self) <*>
+                   Ref.make[Pull[R, E, _]](Pull.end) <*>
+                   Ref.make[Exit[_, _] => URIO[R, _]](_ => UIO.unit) <*>
+                   Ref.make[List[(Pull[R, E, _], Exit[_, _] => URIO[R, _], Structure[R, E, _])]](Nil)).toManaged_
+          (currStruct <*> currPull <*> currFin <*> stack) = refs
+          pull <- ZManaged.reserve(
+                   Reservation(
+                     open(self, currFin, currPull).as(pull(currStruct, currFin, currPull, stack)),
+                     e =>
+                       for {
+                         fs    <- currFin.get.zipWith(stack.get.map(_.map(_._2)))(_ :: _)
+                         exits <- ZIO.foreach(fs)(_(e).run)
+                         _     <- ZIO.done(Exit.collectAll(exits).getOrElse(Exit.unit))
+                       } yield ()
+                   )
+                 )
+        } yield pull
+    }
+
+  private[this] def open(
+    struct: Structure[R, E, Any],
+    currFin: Ref[Exit[_, _] => URIO[R, _]],
+    currPull: Ref[Pull[R, E, _]]
+  ): ZIO[R, E, Unit] =
+    ZIO.uninterruptibleMask { restore =>
+      val proc = struct match {
+        case Iterator(iter) => iter
+        case FlatMap(fa, _) => fa.process
+        case Effect(zio)    => ZManaged.succeed(zio.mapError(Some(_)))
+        case Concat(hd, _)  => hd.process
+      }
+
+      for {
+        reservation <- proc.reserve
+        _           <- currFin.set(reservation.release)
+        _           <- restore(reservation.acquire.tap(currPull.set))
+      } yield ()
+    }
+
+  private[this] def pull(
+    currStruct: Ref[Structure[R, E, Any]],
+    currFin: Ref[Exit[_, _] => URIO[R, _]],
+    currPull: Ref[Pull[R, E, Any]],
+    stack: Ref[List[(Pull[R, E, _], Exit[_, _] => URIO[R, _], Structure[R, E, _])]]
```

Comment:
I feel like some type aliases for finalizers and stack frames would improve readability here.